lib/syntax.ml

```diff
@@ -863,10 +866,138 @@ and eval_nested env nested =
   | Some (t,l) -> join env (resolve_source env t, List.map (fun (x,jt,jc) -> resolve_source env x, jt, jc) l)
   | None -> env, []
 
+(** Extract (sources, name) pairs for columns with IS NOT NULL in WHERE/HAVING
+
+    Handles:
+    - Direct: name IS NOT NULL
+    - Double negation: NOT (name IS NULL)
+    - AND conjunctions: all checks must pass
+    - OR disjunctions: only if ALL branches have the same check
+    - De Morgan's law: NOT (A OR B) extracts from B if B is IS NULL check
+    - Choices: only if ALL branches have the check
```

Comment:
can add NOT (A AND B) as well?
also it should be possible to track this better by bubbling up `Is_null | Is_not_null | Cannot_say` for each expr and then applying negation and logic rules uniformly instead of elaborating each cases by case?